bgpd/bgp_network.c

```diff
@@ -519,7 +519,7 @@ static void bgp_accept(struct event *thread)
 	}
 
 	if (!peer1) {
-		if (bgp_debug_neighbor_events(NULL)) {
+		if (bgp_debug_neighbor_events(NULL) && bgp) {
```

Comment:
it looks as if `bgp` has already been used before this point - do you think this is possible?

Reply:
When listener->bgp is null, it calls bgp_get_instance_for_inc_conn(bgp_sock, &bgp), and when runs into
```
if (!strlen(name)) {
		*bgp_inst = bgp_get_default();
		return 0; /* default instance. */
	}
```

if bgp_get_default() returns null, the return value is 0 and *bgp_inst will be null as well.

Reply:
Do you know how to replicate and hit this code practically?

Reply:
No, I don't know how to hit it. But I have a backtrace from my customer, which shows that it crash at the position that I changed. sigaction also reports si_addr=0x1c, which is the address of bgp->vrf_id when bgp is null.

Reply:
Can we get at least the configuration in what cases this is happening? (I assume this is with netns).

Reply:
There are 2 bgp view instances and no bgp instance in global vrf.  Is this the reason that bgp_get_default() return null?
```
router bgp 65175 view north
  no bgp ebgp-requires-policy
  no bgp default ipv4-unicast
  bgp bestpath as-path multipath-relax
  neighbor 10.156.62.19 remote-as 65108
  neighbor 10.156.62.19 bfd
  neighbor 10.156.62.19 solo
  neighbor 10.156.62.19 ebgp-multihop 255
  neighbor 10.156.62.19 update-source 172.28.219.11
  neighbor 10.156.62.19 timers connect 2
  neighbor 10.156.62.20 remote-as 65108
  neighbor 10.156.62.20 bfd
  neighbor 10.156.62.20 solo
  neighbor 10.156.62.20 ebgp-multihop 255
  neighbor 10.156.62.20 update-source 172.28.219.11
  neighbor 10.156.62.20 timers connect 2
  neighbor 2008:2888:0:f:0:28:0:1 remote-as 65108
  neighbor 2008:2888:0:f:0:28:0:1 bfd
  neighbor 2008:2888:0:f:0:28:0:1 solo
  neighbor 2008:2888:0:f:0:28:0:1 ebgp-multihop 255
  neighbor 2008:2888:0:f:0:28:0:1 update-source 2706:f160:a:9434:0:a09:0:11
  neighbor 2008:2888:0:f:0:28:0:1 timers connect 2
  neighbor 2008:2888:0:f:0:28:0:2 remote-as 65108
  neighbor 2008:2888:0:f:0:28:0:2 bfd
  neighbor 2008:2888:0:f:0:28:0:2 solo
  neighbor 2008:2888:0:f:0:28:0:2 ebgp-multihop 255
  neighbor 2008:2888:0:f:0:28:0:2 update-source 2706:f160:a:9434:0:a09:0:11
  neighbor 2008:2888:0:f:0:28:0:2 timers connect 2
  !
  address-family ipv4 unicast
    neighbor 10.156.62.19 activate
    neighbor 10.156.62.19 soft-reconfiguration inbound
    neighbor 10.156.62.20 activate
    neighbor 10.156.62.20 soft-reconfiguration inbound
  exit-address-family
  !
  address-family ipv6 unicast
    neighbor 2008:2888:0:f:0:28:0:1 activate
    neighbor 2008:2888:0:f:0:28:0:1 soft-reconfiguration inbound
    neighbor 2008:2888:0:f:0:28:0:2 activate
    neighbor 2008:2888:0:f:0:28:0:2 soft-reconfiguration inbound
  exit-address-family
  exit
!
router bgp 65175 view south
  no bgp ebgp-requires-policy
  no bgp default ipv4-unicast
  bgp bestpath as-path multipath-relax
  neighbor 2008:2888:a00:f:0:28:0:1 remote-as 65108
  neighbor 2008:2888:a00:f:0:28:0:1 bfd
  neighbor 2008:2888:a00:f:0:28:0:1 solo
  neighbor 2008:2888:a00:f:0:28:0:1 ebgp-multihop 255
  neighbor 2008:2888:a00:f:0:28:0:1 update-source 2706:f160:a:7831::1000:6b
  neighbor 2008:2888:a00:f:0:28:0:1 timers connect 2
  neighbor 2008:2888:a00:f:0:28:0:2 remote-as 65108
  neighbor 2008:2888:a00:f:0:28:0:2 bfd
  neighbor 2008:2888:a00:f:0:28:0:2 solo
  neighbor 2008:2888:a00:f:0:28:0:2 ebgp-multihop 255
  neighbor 2008:2888:a00:f:0:28:0:2 update-source 2706:f160:a:7831::1000:6b
  neighbor 2008:2888:a00:f:0:28:0:2 timers connect 2
  !
  address-family ipv6 unicast
    neighbor 2008:2888:a00:f:0:28:0:1 activate
    neighbor 2008:2888:a00:f:0:28:0:1 soft-reconfiguration inbound
    neighbor 2008:2888:a00:f:0:28:0:2 activate
    neighbor 2008:2888:a00:f:0:28:0:2 soft-reconfiguration inbound
  exit-address-family
  exit
```

Reply:
why isn't this the fix then:

```
diff --git a/bgpd/bgp_network.c b/bgpd/bgp_network.c
index de57d91806..32f862927b 100644
--- a/bgpd/bgp_network.c
+++ b/bgpd/bgp_network.c
@@ -303,7 +303,8 @@ static int bgp_get_instance_for_inc_conn(int sock, struct bgp **bgp_inst)
 
        if (!strlen(name)) {
                *bgp_inst = bgp_get_default();
-               return 0; /* default instance. */
+
+               return !*bgp_inst; /* default instance. */
        }
 
        /* First try match to instance; if that fails, check for interfaces. */
```

Reply:
although I suspect this will stop the crash but not actually allow the connection from going through which is not what is really wanted.

I would prefer this:

```
diff --git a/bgpd/bgp_network.c b/bgpd/bgp_network.c
index de57d91806..b41bc27044 100644
--- a/bgpd/bgp_network.c
+++ b/bgpd/bgp_network.c
@@ -520,10 +520,14 @@ static void bgp_accept(struct event *thread)
 
        if (!peer1) {
                if (bgp_debug_neighbor_events(NULL)) {
-                       zlog_debug(
-                               "[Event] %s connection rejected(%s:%u:%s) - not configured and not valid for dynamic",
-                               inet_sutop(&su, buf), bgp->name_pretty, bgp->as,
-                               VRF_LOGNAME(vrf_lookup_by_id(bgp->vrf_id)));
+                       if (bgp)
+                               zlog_debug(
+                                       "[Event] %s connection rejected(%s:%u:%s) - not configured and not valid for dynamic",
+                                       inet_sutop(&su, buf), bgp->name_pretty, bgp->as,
+                                       VRF_LOGNAME(vrf_lookup_by_id(bgp->vrf_id)));
+                       else
+                               zlog_debug("[Event] %s connection rejected - not configured and not valid for dynamic",
+                                          inet_sutop(&su, buf));
                }
                close(bgp_sock);
                return;
```

Reply:
Yes, you have better fix!

Reply:
@tufeigunchu -> please modify your PR with this new change and I'll get it in.

Reply:
Done